src/AbstractValueParser.php

```diff
@@ -28,7 +41,9 @@ public function defaultsTo($defaultValue)
 
     public function required()
     {
-        if (is_null($this->value)) {
+        if ($this->invalid) {
+            throw $this->createInvalidValueException();
+        } elseif (is_null($this->value)) {
```

Comment:
Yep, `$value === null && $invalid === false` can be true when the value is not set at all. For example for `$this->queryParameter('value')->int()->required()` with the request `GET /?value=a`, then `$value` is `null` and `$invalid` is `true`. But for `GET /` `$value` is `null` *and* `$invalid` is false, because the value parameter is not passed at all (it can't be invalid because it was not even provided).